gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,141 @@
+#include <dirent.h>
+#include <sys/stat.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
 
 PG_MODULE_MAGIC;
+
+#define should_skip(filename) \
+	(pg_strncasecmp(filename, "pg", 2) == 0 \
+	 || pg_strncasecmp(filename, "t_", 2) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 4, "_fsm", 4) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 3, "_vm", 3) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 5, "_init", 5) == 0)
```

Comment:
What if `strlen(filename) == 1`?

Reply:
When using `pg_strncasecmp`, you should always check the length of the string first, and if the length of the string is less than the pattern then skip the pointless call to `pg_strncasecmp`.

Reply:
And in general, because since all patterns are known in advance, I would first calculate the length of the string, and then compare the first or last few characters, character by character, thus avoiding calling the `pg_strncasecmp` function altogether.